File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/stateengine/ParallelStack.java
package org.firstinspires.ftc.teamcode.teamcode.stateengine;

import org.firstinspires.ftc.teamcode.teamcode.hardware.RobotHardware;

public class ParallelStack extends StackList implements State {

    private boolean isDone = false;

    @Override
    public void init(RobotHardware rh) {
        super.initStack(rh);
    }

    @Override
    public void run() {
        boolean done = true;

        for (State state : getStack()) {
            if (!state.isDone()) {
                state.run();
                done = false;
            }
        }

        // when all of the states are done,
        // this stack will be done
        if (done) {
            isDone = true;
        }
    }

    @Override
    public void stop() {

    }

    @Override
    public boolean isDone() {
        return isDone;
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/stateengine/ParallelStack.kt
package org.firstinspires.ftc.teamcode.stateengine

import org.firstinspires.ftc.teamcode.hardware.RobotHardware

class ParallelStack(private val stack: Array<State>) : StackList(stack), State {
  override var isDone: Boolean = false

  override fun init(rh: RobotHardware) {
    super.initStack(rh)
  }

  override fun run() {
    if (this.isDone) return
    var done = true

    for (state in stack) {
      if (!state.isDone) {
        state.run()
        done = false
      }
    }

    // When all of the states are done, this stack will be done
    if (done) {
      this.isDone = true
    }
  }
}
